### PR TITLE
fix: remote srcset images not being resized and deduplication not working in certain cases

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

Continuation of https://github.com/withastro/astro/pull/8819

This PR also reworks completely how we do the transforms for `srcset`, putting special care into avoiding regenerating images when unnecessary. It's possible that there's still some edge cases where it doesn't work, nonetheless as far as I could tell, it does work!

This also fixes https://github.com/withastro/astro/issues/8839, since it's in the same area.

## Testing

Added tests!

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/assets/services/service.ts [hunk 1]
```diff
@@ -245,62 +255,46 @@ export const baseService: Omit<LocalImageService, 'transform'> = {
...
-					srcSetValue.transform.width = maxTargetWidth;
-					srcSetValue.transform.height = Math.round(maxTargetWidth / aspectRatio);
-				}
+			allWidths.push(...widths.map((width) => ({ width, descriptor: `${width}w` as const })));
+		}
 
-				if (targetFormat !== options.format) {
-					srcSetValue.transform.format = targetFormat;
+		// Caution: The logic below is a bit tricky, as we need to make sure we don't generate the same image multiple times
+		// When making changes, make sure to test with different combinations of local/remote images widths, densities, and dimensions etc.
```
Comment (Princesseuh) [inline packages/astro/src/assets/services/service.ts:269]: Added a "There be dragons" comment here because it's honestly super tricky to completely avoid making transforms that are functionally equivalent to another one, but don't have quite the exact same shape.

Common examples includes:
- A transform where `webp` is specified manually vs one where it's not. Due to the default, both of those results in the same image.
- A transform where `width` is specified, but it's equivalent to the width of the image, so it's not resized.

File: packages/astro/src/assets/utils/transformToPath.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import { deterministicString } from 'deterministic-object-hash';
```
Comment (Princesseuh) [inline packages/astro/src/assets/utils/transformToPath.ts:1]: I found this package in the backwoods somewhere in a random stack overflow thread, but it's honestly quite nice and does everything I wanted.

**Comment by Princesseuh** [comment]
!preview srcset-remote

**Comment by itsmatteomanf** [comment]
Is there *any* chance to know when is this gonna get merged? It's a major bug fix on the newly announced feature...

I know this issue exists so I won't use `srcset` yet, but I'd really like to ahah

**Comment by Princesseuh** [comment]
> Is there *any* chance to know when is this gonna get merged? It's a major bug fix on the newly announced feature...
> 
> I know this issue exists so I won't use `srcset` yet, but I'd really like to ahah

We'll release it Monday morning CET, it needs to get reviewed and we don't typically do releases on the weekend

File: packages/astro/src/assets/services/service.ts [hunk 3]
```diff
@@ -245,62 +259,46 @@ export const baseService: Omit<LocalImageService, 'transform'> = {
...
+				if (options.width && options.height) {
+					transform.width = options.width;
+					transform.height = options.height;
 				}
+			}
 
-				srcSet.push(srcSetValue);
+			srcSet.push({
+				transform: transform,
+				descriptor: descriptor,
```
Comment (ematipico) [inline packages/astro/src/assets/services/service.ts:296]:
```suggestion
				transform,
				descriptor,
```

**Review by ematipico — APPROVED** [review]
The logic is very complex, I can't really understand it.

**Comment by ematipico** [comment]
It's okay, you did a great job! Thank you for addressing my comments. Feel free to merge it whenever you want. I don't have big concerns
